File: src/types/ytdlp.ts
/** YouTube動画のメタ情報 */
export interface VideoInfo {
  id: string;
  title: string;
  description: string;
  channel: string;
  channelUrl: string;
  duration: number; // 秒
  uploadDate: string; // "20260320"
  viewCount: number;
  thumbnail: string; // URL
  thumbnails: Thumbnail[];
  formats: DownloadFormat[];
  requestedSubtitles?: Record<string, SubtitleInfo>;
}

export interface Thumbnail {
  url: string;
  width: number;
  height: number;
}

/** ダウンロードフォーマット */
export interface DownloadFormat {
  formatId: string;
  formatNote: string;
  ext: string;
  resolution?: string;
  fps?: number;
  vcodec?: string;
  acodec?: string;
  videoBitrate?: number;
  audioBitrate?: number;
  filesize?: number;
  filesizeApprox?: number;
  hasVideo: boolean;
  hasAudio: boolean;
}

export interface SubtitleInfo {
  ext: string;
  url: string;
  name: string;
}

/** ダウンロードパラメータ */
export interface DownloadParams {
  url: string;
  formatId: string;
  outputDir: string;
  filename?: string;
  mergeFormat?: string;
}

/** ダウンロード進捗 */
export interface DownloadProgress {
  percent: number;
  downloadedBytes: number;
  totalBytes?: number;
  speed?: string;
  eta?: string;
  status: DownloadStatus;
}

export type DownloadStatus =
  | 'downloading'
  | 'merging'
  | 'post-processing'
  | 'complete'
  | 'error';

File: src-tauri/src/models/ytdlp.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct VideoInfo {
    pub id: String,
    pub title: String,
    pub description: String,
    pub channel: String,
    pub channel_url: String,
    pub duration: f64,
    pub upload_date: String,
    pub view_count: u64,
    pub thumbnail: String,
    pub thumbnails: Vec<Thumbnail>,
    pub formats: Vec<DownloadFormat>,
    pub requested_subtitles: Option<HashMap<String, SubtitleInfo>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Thumbnail {
    pub url: String,
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DownloadFormat {
    pub format_id: String,
    pub format_note: String,
    pub ext: String,
    pub resolution: Option<String>,
    pub fps: Option<f64>,
    pub vcodec: Option<String>,
    pub acodec: Option<String>,
    pub video_bitrate: Option<f64>,
    pub audio_bitrate: Option<f64>,
    pub filesize: Option<u64>,
    pub filesize_approx: Option<u64>,
    pub has_video: bool,
    pub has_audio: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SubtitleInfo {
    pub ext: String,
    pub url: String,
    pub name: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DownloadProgress {
    pub percent: f64,
    pub downloaded_bytes: u64,
    pub total_bytes: Option<u64>,
    pub speed: Option<String>,
    pub eta: Option<String>,
    pub status: String,
}
